websearch/websearch/src/durability.rs

```diff
@@ -0,0 +1,579 @@
+                        partial_results,
+                        finished,
+                        ..
+                    }) => {
+                        if *finished {
+                            (Ok(Vec::new()), None)
+                        } else {
+                            let retry_params = Impl::retry_params(original_params, partial_results);
+
+                            let (session, first_live_result) =
+                                with_persistence_level(PersistenceLevel::PersistNothing, || {
+                                    let session =
+                                        <Impl as ExtendedwebsearchGuest>::unwrapped_search_session(
+                                            retry_params,
+                                        )
+                                        .unwrap();
+
+                                    for lazy_initialized_pollable in pollables {
+                                        lazy_initialized_pollable.set(Impl::subscribe(&session));
+                                    }
+
+                                    let next = session.next_page();
+                                    (session, next)
+                                });
+                            let value = first_live_result.clone()?;
+                            let _ = durability.persist_infallible(NoInput, value);
+
+                            (first_live_result, Some(session))
+                        }
+                    }
+                    None => {
+                        unreachable!()
+                    }
+                };
+
+                if let Some(session) = new_live_session {
+                    let pollables = match state.take() {
+                        Some(DurableSearchSessionState::Live { pollables, .. }) => pollables,
+                        Some(DurableSearchSessionState::Replay { pollables, .. }) => pollables,
+                        None => {
+                            unreachable!()
+                        }
+                    };
+                    *state = Some(DurableSearchSessionState::Live { session, pollables });
+                }
+
+                result
+            } else {
+                let result: Vec<SearchResult> = durability.replay_infallible();
+                let mut state = self.state.borrow_mut();
+                match &mut *state {
+                    Some(DurableSearchSessionState::Live { .. }) => {
+                        unreachable!("Durable search session cannot be in live mode during replay");
+                    }
+                    Some(DurableSearchSessionState::Replay {
+                        partial_results: _,
+                        finished: _,
+                        ..
+                    }) => Ok(result),
+                    None => {
+                        unreachable!();
+                    }
+                }
+            }
+        }
+
+        fn get_metadata(&self) -> Option<SearchMetadata> {
```

Comment:
optional: I would not make getting the metadata durable, but rather just ensure that it can be deterministically computed based on the previous page (potentially as a provider specific type)